share/source.F90

```diff
@@ -12,35 +12,58 @@
 ! Full instructions on compiling and using MCNP5 with this subroutine are found
 ! in the PyNE user manual.
 
-function find_cell() result(icl_tmp)
-! This function to determines the current MCNP cell index location and exits if
-! no valid cell is found. This only works if there are no repeated geometries or
-! universes present in the model.
+function find_cell(cell_list, cell_list_size) result(icl_tmp)
+! This function to determines the current MCNP cell index location.
+! Return a positive integer if a valid cell is found, otherwise it returns -1.
+! This only works if there are no repeated geometries or universes present in
+! the model.
+! Parameters:
+!     cell_list: array of integers. Contains cell numbers that the source
+!                particle possiblely located.
+!     cell_list_size: integer. Size of the cell_list.
 
     use mcnp_global
     use mcnp_debug
     ! xxx,yyy,zzz are global variables
     ! mxa is global
     integer :: i ! iterator variable
-    integer :: j ! tempory cell test
+    integer :: j ! temporary cell test
     integer :: icl_tmp ! temporary cell variable
-    icl_tmp = -1
-
-    do i = 1, mxa
-      call chkcel(i, 0, j)
-      if (j .eq. 0) then
-         ! valid cel set
-         icl_tmp = i
-         exit
-      endif
-    enddo
-    ! icl is now set
+    integer, intent(in) :: cell_list_size
+    integer, dimension(cell_list_size), intent(in) :: cell_list
+    integer :: cid ! cell index
 
-    if(icl_tmp .le. 0) then
-      write(*,*) 'Nonsense cell number stopping'
-      stop
+    icl_tmp = -1
+    if (cell_list_size .eq. 0) then
+        ! TET mesh
+        do i = 1, mxa
+           call chkcel(i, 0, j)
+           if (j .eq. 0) then
+              ! valid cell found
+              icl_tmp = i
+              exit
+           endif
+        enddo
+    else
+        ! HEX mesh. VOXEL/SUBVOXEL
+        do i = 1, cell_list_size
+           if (cell_list(i) .le. 0) then
```

Comment:
If we know the number of valid entries when we pass the `cell_list` to `find_cell` we should just pass that value as `cell_list_size`. This check can then be removed.

Reply:
We can get the valid cell list size by two options:
- A: pass back another variable
- B: find the first negative value in the cell list

But I both these two options just remove the check here, but add checks at other places. The differences are the speed of the different options. I think we should consider that after we are sure the correctness of the PR.

Reply:
I'm confused... can you clarify?

What do you mean by pass back another variable? 

What checks would be added in other places?

Reply:
For the pass back another variable part. We have to pass back the `cell_list_size` because we need to allocate the `cell_list`. And now we want to know the number of valid cells, we can pass pack another variable named such as `valid_num_cells` or `valid_cell_list_size`.

Reply:
For checks at other places. If we don't pass back an extra variable of the valid cell list size, then we have to calculate it with the given `cell_list`. This calculation will performed with code blocks iteration and check of -1, or something like find index of first -1.

Reply:
Ah yeah I see what you mean because.  We don't know how many valid cells there are when we're building the `cell_list` because the tag is fixed length and padded with -1 entries... and yeah that ties into a check somewhere else for that first -1 entry. Rats.